src/main/java/TicTacToe.java

```diff
@@ -0,0 +1,327 @@
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+// Strategy pattern.
+// The computer player's behavior/strategy can be replaced by inheriting from the interface below
+// Also, the human player's behavior inherits from the same interface
+// This also makes it easy to modify the game for 2 human players, 2 computer players etc.
+
+interface MoveMethod
+{
+    public int move();
+}
+
+class SimpleMoveStrategy implements MoveMethod
+{
+    private TicTacToe game;
+    public SimpleMoveStrategy(TicTacToe t) {
+        game = t ;
+    }
+
+    public int move() {
+
+        for( int i = 0 ; i < TicTacToe.N ; i ++ )
+            {
+                for( int j = 0 ; j < TicTacToe.N ; j ++ )
+                    {
+                        if( game.board[i][j] == 0 )
+                            return (i*3+j+1) ;
+                    }
+            }
+        return 0 ;
+    }
+}
+
+class HumanMove implements MoveMethod
+{
+    private TicTacToe game;
+    public HumanMove(TicTacToe t) {
+        game = t ;
+    }
+    public int move() {
+
+        String move_str ;
+        int move_int = 0 ;
+        boolean valid_input = false ;
+        while(!valid_input) {
+            System.out.print("Where to ? ");
+            move_str = TicTacToe.getUserInput() ;
+            if( Character.isDigit(move_str.toCharArray()[0]) ) {
+                move_int = Integer.parseInt(move_str);
+                if( ( move_int <= (TicTacToe.N)*(TicTacToe.N) ) && move_int >= 1 ) {
+                    valid_input = true ;
+                    break ;
+                }
+            }
+
+            if( !valid_input ) {
+                System.out.println("Invalid input");
+                continue ;
+            }
+        }
+        return move_int ;
+    }
+
+}
+
+class TicTacToe
+{
+    protected static final int N = 3 ;
+    private static final int HSPACE = 20 ;
+    protected int[][] board;
+    private static BufferedReader reader =
+        new BufferedReader(new InputStreamReader(System.in)) ;
+
+    class Player
+    {
+        private String name ;
+        private int player_type ;
+        private int player_order ;
+        private MoveMethod move_strategy ;
+
+        public Player(String pname, int type, int order, MoveMethod move_s )

```

Comment:
pname -> name (clearer)
move_s  -> move_strategy   (keep same name as the local for clarity)